### ISSUE TITLE
[503] "Tasks that run when changed should likely be handlers" is too opinionated and annoying

### ISSUE THREAD

**Issue by geerlingguy** [ansible-community/ansible-lint#419]
# Issue Type

- Bug report

# Ansible and Ansible Lint details

```
$ ansible --version
ansible 2.7.4
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/Users/jeff.geerling/.ansible/plugins/modules',u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.15 (default, Jul 23 2018, 21:27:06) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]

$ ansible-lint --version
ansible-lint 3.5.1
```

- ansible installation method: pip
- ansible-lint installation method: pip

# Desired Behaviour

Reacting to a change (e.g. `when: something.changed`) is useful for many scenarios where using a handler is way too heavyweight.

# Actual Behaviour (Bug report only)

If I do something like "clone a repo, if it was changed, do xyz before doing other things", then I get the following report:

```
[503] Tasks that run when changed should likely be handlers
/Users/jeff.geerling/Dropbox/VMs/roles/geerlingguy.drush/tasks/install-source.yml:17
Task/Handler: Ensure Drush can be installed on Debian Wheezy.

[503] Tasks that run when changed should likely be handlers
/Users/jeff.geerling/Dropbox/VMs/roles/geerlingguy.drush/tasks/install-source.yml:26
Task/Handler: Install Drush dependencies with Composer.
```

Here's that example code:

```yaml
- name: Clone Drush from GitHub.
  git:
    repo: https://github.com/drush-ops/drush.git
    dest: "{{ drush_source_install_path }}"
    version: "{{ drush_source_install_version }}"
    update: "{{ drush_keep_updated }}"
    force: "{{ drush_force_update }}"
    depth: "{{ drush_clone_depth }}"
  register: drush_clone

- name: Check for composer.json
  stat: path={{ drush_source_install_path }}/composer.json
  register: drush_composer

# See: https://github.com/geerlingguy/ansible-role-drush/issues/6
- name: Ensure Drush can be installed on Debian Wheezy.
  command: >
    {{ composer_path }} update {{ drush_composer_cli_options }}
    chdir={{ drush_source_install_path }}
  when:
    - drush_clone.changed and ansible_distribution == "Debian"
    - ansible_distribution_release == "wheezy"
    - drush_composer.stat.exists

- name: Install Drush dependencies with Composer.
  command: >
    {{ composer_path }} install {{ drush_composer_cli_options }}
    chdir={{ drush_source_install_path }}
  when: (drush_clone.changed and drush_composer.stat.exists) or drush_force_composer_install
```

Actual code: https://github.com/geerlingguy/ansible-role-drush/blob/master/tasks/install-source.yml#L2-L25

I have rewritten these tasks to use handlers, and it ends up being an extra 5-10 lines of code overall, and it makes the flow of the tasks way more complex... plus I have to do a `meta: flush_handlers` inline, since I need the follow-up tasks to execute _now_ before other parts of my role (which could be affected by the new code) have a chance to run.

I think having to call `meta: flush_handlers` can also be slightly dangerous (and IMO is a code smell) because you might be calling a bunch of other handlers if nothing else has flushed them by the time this particular role runs.

While I think this is a warning and not an 'error', it still dings my module's 'quality score' on Ansible Galaxy, and also fails my builds in Travis CI (exit code is `2` and not `0`), so I would vote for dropping it entirely, or finding a way to report it as more a general info thing (instead of exiting with a non-0 failure).

**Comment by geerlingguy** [comment]
Currently I'm using `tags: ['skip_ansible_lint'] for all these instances, and I have had to use this hammer over 30 times through my 100+ roles to squelch the error. I hate doing that, because that means `ansible-lint` skips that entire task and doesn't just skip the 503 lint rule.

**Comment by gundalow** [comment]
If we don't revisit 503, maybe adding the ability to skip a specific rule is needed.

**Comment by greg-hellings** [comment]
I, for one, like this behavior. I absolutely want `ansible-lint` to trigger on this behavior. However, having the ability to ignore/disable only certain rules would be very useful as @gundalow suggests.

**Comment by geerlingguy** [comment]
Apparently there might already be a way to skip a rule for an entire role:

```
skip_list:
  - '503'
```

Found on #451.

**Comment by codestardustco** [comment]
+1, sometimes processes need to be run in the middle of the play (such as a restart to update configs) and handlers aren't designed for that.

**Comment by stemid** [comment]
Am I using an anti-pattern? Because I'm often doing things in this order, either for deploy or update of deploy.

1. git update
2. when git update|changed I do a python setup.py install command

And this of course triggers the lint error.

I can't use a handler instead because the command module is dependent on values I get from a role.

**Comment by willthames** [comment]
@stemid yeah, using ansible to run commands is pretty much an anti pattern (it's very difficult to get it right: https://willthames.github.io/2016/09/21/using-command-and-shell-in-ansible.html)

Ideally step two would be an idempotent module and then you don't need the `when: changed`. I know that's not always possible, but that's when you want to use the `skip_ansible_lint` tag

**Comment by geerlingguy** [comment]
> Ideally step two would be an idempotent module and then you don't need the `when: changed`. I know that's not always possible, but that's when you want to use the `skip_ansible_lint` tag

The problem is, I think 99% of the time I don't want to spend the extra hour or two to write a custom Python module to manage a service I'm quickly wrapping in an Ansible role or playbook.

And adding `skip_ansible_lint` means that entire task is skipped, when it could have other issues that the linter could pick up besides 503.

And you could add `#noqa 503` to all these kinds of tasks, but that's also burdensome IMO.

I've literally added a 503 exclude rule to every one of my repositories now... I just think (as I initially stated in the OP) that this rule is overly opinionated and in real world Ansible usage, there are thousands of situations where it's much better to use `when: something is changed` instead of writing a new module or trying to do a dance with `handlers` and `meta` manipulation (which can result in much more surprising and/or damaging behavior).

**Comment by greg-hellings** [comment]
I've been adopting the `#noqa 503` method, personally. I still like to be notified, particularly on PRs from other developer whose judgement I don't yet trust, that this generalize rule has been broken. It also at least makes me pause to reconsider when I use it, myself, if it's really justified.

Is it annoying? Yeah. But all linters are, in my experience! My code is usually better for them, though.

**Comment by ericsysmin** [comment]
I ran into this one as well. Seriously though, this one shouldn't be an error, maybe a non-blocking warning at most

**Comment by ssbarnea** [comment]
I am closing it as no PR was made and because disabling it would upset a considerable number of users that want to know about possible problems with their code.

Unless someone comes with another brilliant idea, I think that those affect can either disable the entire rule, make it a warning or use localized noqa.

**Comment by phgogo** [comment]
In my opinion, E503 should be splitted into two rules, that can disabled/skipped independently.

One rule for a single `when: xyz.changed` 
and another rule for more complex conditions like `when: xyz.changed and weatherIsNice|bool`

While i would highly recommend to use handlers in the first example, i can think of situations in which the second example might be solved better without a handler (as multiple others have pointed out in this tread)

**Comment by ssbarnea** [comment]
There is no need for a split in this case because it would make sense to just make the rule spot basic conditions and not fire on complex ones. I will lock this issue because it is very old and many thing changed since, include the rule id and I do not want to spam all involved parties. Feel free to raise a PR and mention this ticket. Keep in mind that there is also the "foo is changed" syntax.
